Greetings, and thanks for all of your work on the wonderful rsync!

I recently had the need to transfer files only with different mod
dates (and to *not* transfer them based on file size differences).
This is because I'm backing up files remotely on an untrusted machine,
so I'm encrypting them with gpg before transfer.  I discovered that
rsync didn't already have a --date-only flag, so I added one and am
enclosing the diffs in case you (as I hope) decide to include this
option in future releases.

Again, thanks!

Best Regards,
Jeremy Bornstein

[Patched update to have context and apply to latest CVS source.]

--- orig/generator.c	2005-08-17 06:45:07
+++ generator.c	2004-11-11 22:15:27
@@ -58,6 +58,7 @@ extern int append_mode;
 extern int make_backups;
 extern int csum_length;
 extern int ignore_times;
+extern int date_only;
 extern int size_only;
 extern OFF_T max_size;
 extern int io_error;
@@ -360,6 +361,8 @@ void itemize(struct file_struct *file, i
 /* Perform our quick-check heuristic for determining if a file is unchanged. */
 static int unchanged_file(char *fn, struct file_struct *file, STRUCT_STAT *st)
 {
+	if (date_only)
+		return cmp_modtime(st->st_mtime, file->modtime) == 0;
 	if (st->st_size != file->length)
 		return 0;
 
--- orig/options.c	2005-08-27 21:11:26
+++ options.c	2005-08-27 21:19:31
@@ -90,6 +90,7 @@ int keep_partial = 0;
 int safe_symlinks = 0;
 int copy_unsafe_links = 0;
 int size_only = 0;
+int date_only = 0;
 int daemon_bwlimit = 0;
 int bwlimit = 0;
 int fuzzy_basis = 0;
@@ -316,6 +317,7 @@ void usage(enum logcode F)
   rprintf(F,"     --timeout=TIME          set I/O timeout in seconds\n");
   rprintf(F," -I, --ignore-times          don't skip files that match in size and mod-time\n");
   rprintf(F,"     --size-only             skip files that match in size\n");
+  rprintf(F,"     --date-only             skip files that match in mod-time\n");
   rprintf(F,"     --modify-window=NUM     compare mod-times with reduced accuracy\n");
   rprintf(F," -T, --temp-dir=DIR          create temporary files in directory DIR\n");
   rprintf(F," -y, --fuzzy                 find similar file for basis if no dest file\n");
@@ -413,6 +415,7 @@ static struct poptOption long_options[] 
   {"no-implied-dirs",  0,  POPT_ARG_VAL,    &implied_dirs, 0, 0, 0 },
   {"ignore-times",    'I', POPT_ARG_NONE,   &ignore_times, 0, 0, 0 },
   {"size-only",        0,  POPT_ARG_NONE,   &size_only, 0, 0, 0 },
+  {"date-only",        0,  POPT_ARG_NONE,   &date_only, 0, 0, 0 },
   {"one-file-system", 'x', POPT_ARG_NONE,   &one_file_system, 0, 0, 0 },
   {"update",          'u', POPT_ARG_NONE,   &update_only, 0, 0, 0 },
   {"existing",         0,  POPT_ARG_NONE,   &only_existing, 0, 0, 0 },
@@ -1445,6 +1448,9 @@ void server_options(char **args,int *arg
 	if (size_only)
 		args[ac++] = "--size-only";
 
+	if (date_only)
+		args[ac++] = "--date-only";
+
 	if (modify_window_set) {
 		if (asprintf(&arg, "--modify-window=%d", modify_window) < 0)
 			goto oom;
--- orig/rsync.yo	2005-08-27 21:05:12
+++ rsync.yo	2005-02-11 22:49:10
@@ -350,6 +350,7 @@ to the detailed description below for a 
      --timeout=TIME          set I/O timeout in seconds
  -I, --ignore-times          don't skip files that match size and time
      --size-only             skip files that match in size
+     --date-only             skip files that match in mod-time
      --modify-window=NUM     compare mod-times with reduced accuracy
  -T, --temp-dir=DIR          create temporary files in directory DIR
  -y, --fuzzy                 find similar file for basis if no dest file
@@ -448,6 +449,12 @@ regardless of timestamp. This is useful 
 after using another mirroring system which may not preserve timestamps
 exactly.
 
+dit(bf(--date-only)) Normally rsync will skip any files that are
+already the same size and have the same modification time-stamp. With the
+--date-only option, files will be skipped if they have the same
+timestamp, regardless of size. This may be useful when the remote
+files have passed through a size-changing filter, e.g. for encryption.
+
 dit(bf(--modify-window)) When comparing two timestamps, rsync treats the
 timestamps as being equal if they differ by no more than the modify-window
 value.  This is normally 0 (for an exact match), but you may find it useful
